### PR TITLE
Read list of courses during authentication, if available.

### PR DESCRIPTION
**Pull request by rocha**

### CODE REVIEW

**Comment by rocha** [comment]
@dsjen @clintonb review please.

I significantly reduced the amount of code going into this PR. Initially it included logic to retrieve only the courses that were accessed, and the list of courses only on the course list page. However, these optimizations were not useful because of the bottleneck in the LMS.

Finally, I had to add a `test_login` test, since it seems that we weren't testing for the good login path, just for login failures.

File: analytics_dashboard/courses/permissions.py [hunk 1]
```diff
@@ -60,7 +63,7 @@ def refresh_user_course_permissions(user):
     Arguments
         user (User) --  User whose permissions should be refreshed
     """
-    backend = EdXOpenIdConnect(strategy=load_strategy())
```
Comment (clintonb) [inline analytics_dashboard/courses/permissions.py:63]: What's the motivation for this change?

File: analytics_dashboard/analytics_dashboard/settings/test.py [hunk 2]
```diff
@@ -21,3 +21,5 @@
 
 ENABLE_AUTO_AUTH = True
 SOCIAL_AUTH_EDX_OIDC_URL_ROOT = 'http://example.com'
+
+COURSE_PERMISSIONS_CLAIMS = ['staff_courses']
```
Comment (clintonb) [inline analytics_dashboard/analytics_dashboard/settings/test.py]: Set `COURSE_PERMISSIONS_CLAIMS` to an empty list in `settings/base.py`. Also, please update the README with any relevant info (or, at a minimum, add a comment in `base.py`).

File: analytics_dashboard/analytics_dashboard/backends.py [hunk 3]
```diff
@@ -32,10 +35,50 @@ class EdXOpenIdConnect(OpenIdConnectAuth):
...
+        locale = response.get('locale')
+        if locale:
+            details[u'language'] = _to_language(response['locale'])
+
+        return details
+
+    def auth_complete_params(self, state=None):
+        params = super(EdXOpenIdConnect, self).auth_complete_params(state)
+
+        # TODO: Due a limitation in the oidc provider in the LMS, the list of all course permissions
```
Comment (dsjen) [inline analytics_dashboard/analytics_dashboard/backends.py:45]: Is there a corresponding task to track this?

File: analytics_dashboard/courses/permissions.py (hunk 1 above)
Comment (rocha) [inline analytics_dashboard/courses/permissions.py:63]: Circular imports. `backends.py` is importing `permissions.py` and vice versa. In the previous version I had a separate module to avoid the circular import, however it still wasn't a good solution.

Any suggestions?

File: analytics_dashboard/analytics_dashboard/backends.py (hunk 3 above)
Comment (rocha) [inline analytics_dashboard/analytics_dashboard/backends.py:45]: No, it is really on the LMS. I'll open a story, however it is likely that this is not going to be addressed for a long time since it requires major changes.

File: analytics_dashboard/analytics_dashboard/settings/test.py (hunk 2 above)
Comment (rocha) [inline analytics_dashboard/analytics_dashboard/settings/test.py]: `COURSE_PERMISSIONS_CLAIMS` is already set to this value in `base.py`. I will remove it from here.

File: analytics_dashboard/analytics_dashboard/backends.py (hunk 3 above)
Comment (dsjen) [inline analytics_dashboard/analytics_dashboard/backends.py:45]: I see.  Thanks!

File: analytics_dashboard/analytics_dashboard/backends.py [hunk 4]
```diff
@@ -32,10 +35,50 @@ class EdXOpenIdConnect(OpenIdConnectAuth):
...
     def user_data(self, _access_token, *_args, **_kwargs):
+        # Include decoded id_token fields in user data.
         return self.id_token
 
+    def get_user_details(self, response):
+        details = self._map_user_details(response)
+
+        locale = response.get('locale')
+        if locale:
+            details[u'language'] = _to_language(response['locale'])
```
Comment (dsjen) [inline analytics_dashboard/analytics_dashboard/backends.py]: Would it make sense to reference `PROFILE_TO_DETAILS_KEY_MAP['locale']`?  Are `details` and `PROFILE_TO_DETAILS_KEY_MAP` related?

File: analytics_dashboard/courses/permissions.py (hunk 1 above)
Comment (clintonb) [inline analytics_dashboard/courses/permissions.py:63]: Ah. I didn't realize that. For safety's sake I'd move the import into `refresh_user_course_permissions` and add a short comment noting the potential for circular import errors.

File: analytics_dashboard/analytics_dashboard/settings/test.py (hunk 2 above)
Comment (clintonb) [inline analytics_dashboard/analytics_dashboard/settings/test.py]: Got it. As long as the value is set to a sane default (which I assume is `[]`) in `base.py`, I'm happy.

File: analytics_dashboard/analytics_dashboard/backends.py (hunk 4 above)
Comment (rocha) [inline analytics_dashboard/analytics_dashboard/backends.py]: Mmm, I didn't add this I think. Let me rebase and figure out where this change came from.

File: analytics_dashboard/courses/permissions.py (hunk 1 above)
Comment (rocha) [inline analytics_dashboard/courses/permissions.py:63]: Good idea.

File: analytics_dashboard/analytics_dashboard/settings/test.py (hunk 2 above)
Comment (rocha) [inline analytics_dashboard/analytics_dashboard/settings/test.py]: Actually the value in `base.py` is the same as in here.

File: analytics_dashboard/analytics_dashboard/backends.py (hunk 4 above)
Comment (rocha) [inline analytics_dashboard/analytics_dashboard/backends.py]: It turns out that what I did is move this chunk of code from one place to another. I restored it to its original location.

Comment (dsjen) [inline analytics_dashboard/analytics_dashboard/backends.py]: Ok!

File: analytics_dashboard/courses/permissions.py [hunk 5]
```diff
@@ -60,7 +62,10 @@ def refresh_user_course_permissions(user):
...
+    # Note that `analtiycs_bashboard.backends` imports the current module
```
Comment (clintonb) [inline analytics_dashboard/courses/permissions.py]: Typo in analytics.

**Comment by clintonb** [comment]
:+1:

File: analytics_dashboard/courses/permissions.py (hunk 5 above)
Comment (rocha) [inline analytics_dashboard/courses/permissions.py]: Fixed.

**Comment by dsjen** [comment]
Looks like there are some build errors.  Looks good to me otherwise!

**Comment by rocha** [comment]
The error is pylint complaining about circular dependencies. I ended up resolving the problem using django signals.

Can you guys check if that looks ok? It looks better than what I had previously involving an intermediate "helper" module.

**Comment by clintonb** [comment]
I'm a bit concerned that this implementation seems a bit overkill given that moving the import statement is a simple and well-known workaround. Other than this concern, :+1:

**Comment by rocha** [comment]
@clintonb I asked a few people and the response what that while the workaround is popular, it is very brittle, and it is clear sign of a design problem. One of the ideas behind Django signals is to break coupling between apps, which I think fits perfectly in this case. I came with another solution previously, but realize that Django had already the same functionality.
